File: projects/DCLD_001/DCLD_001_upper_ui_flutter/lib/core/protocol/dcld_protocol.dart
import 'dart:convert';
import 'dart:typed_data';

import '../models/dcld_models.dart';
import '../version/version_capabilities.dart';

class NormalizedTelemetry {
  const NormalizedTelemetry({required this.telemetry, required this.raw});

  final TelemetryState telemetry;
  final Map<String, dynamic> raw;
}

Map<String, dynamic> buildGetStatusCommand() => {'cmd': 'get_status'};

Map<String, dynamic> buildSetModeCommand(WorkMode mode) => {
  'cmd': 'set_mode',
  'mode': mode.key,
};

Map<String, dynamic> buildSetThresholdCommand(num value) => {
  'cmd': 'set_threshold',
  'value': value,
};

TelemetryState createDefaultTelemetry(VersionCapabilities capabilities) {
  final values = <String, num?>{
    'distance_cm': 0,
    'raw_distance_cm': 0,
    'temperature_c': capabilities.has('temp') ? 20 : null,
    'threshold_cm': 80,
  };

  for (final metric in getSupportedMetrics(capabilities)) {
    values.putIfAbsent(metric.key, () => null);
  }

  return TelemetryState(values: values, mode: WorkMode.auto, alarm: 0, cameraUrl: '');
}

ThresholdState createDefaultThresholds() {
  return ThresholdState({
    for (final threshold in getSupportedThresholds()) threshold.key: threshold.defaultValue,
  });
}

Map<String, dynamic> parseTelemetry(Object? payload) {
  if (payload == null) {
    return <String, dynamic>{};
  }
  if (payload is Map) {
    return Map<String, dynamic>.from(payload);
  }

  final text = payload is Uint8List
      ? utf8.decode(payload)
      : payload is ByteBuffer
          ? utf8.decode(payload.asUint8List())
          : payload.toString();
  try {
    final parsed = jsonDecode(text);
    if (parsed is Map) {
      return Map<String, dynamic>.from(parsed);
    }
    return {'raw': parsed};
  } catch (_) {
    return {'raw': text};
  }
}

NormalizedTelemetry normalizeTelemetry(
  Object? payload,
  VersionCapabilities capabilities, {
  TelemetryState? previous,
  ThresholdState? thresholds,
}) {
  final raw = parseTelemetry(payload);
  final base = previous ?? createDefaultTelemetry(capabilities);
  final thresholdBase = thresholds ?? createDefaultThresholds();
  final nextValues = Map<String, num?>.from(base.values);

  for (final metric in getSupportedMetrics(capabilities)) {
    if (raw.containsKey(metric.key)) {
      nextValues[metric.key] = _toNumber(raw[metric.key], base[metric.key] ?? 0);
    }
  }

  if (raw.containsKey('threshold_cm')) {
    nextValues['threshold_cm'] = _toNumber(raw['threshold_cm'], base['threshold_cm'] ?? 80);
  }

  final nextMode = raw.containsKey('mode') ? WorkMode.fromKey(raw['mode']) : base.mode;
  final nextCameraUrl = raw.containsKey('camera_url') ? (raw['camera_url']?.toString() ?? '') : base.cameraUrl;
  final next = TelemetryState(
    values: nextValues,
    mode: nextMode,
    alarm: raw.containsKey('alarm')
        ? _toSwitch(raw['alarm'])
        : calculateAlarm(
            TelemetryState(values: nextValues, mode: nextMode, alarm: base.alarm, cameraUrl: nextCameraUrl),
            thresholdBase,
          ),
    cameraUrl: nextCameraUrl,
  );

  return NormalizedTelemetry(telemetry: next, raw: raw);
}

ThresholdState normalizeThresholds(Object? payload, {ThresholdState? previous}) {
  final raw = parseTelemetry(payload);
  final source = raw['thresholds'] is Map
      ? Map<String, dynamic>.from(raw['thresholds'] as Map)
      : raw;
  final base = previous ?? createDefaultThresholds();
  final next = Map<String, num>.from(base.values);

  if (source.containsKey('distance')) {
    next['distance'] = _toNumber(source['distance'], base['distance'] ?? 80);
  }
  if (raw.containsKey('threshold_cm')) {
    next['distance'] = _toNumber(raw['threshold_cm'], base['distance'] ?? 80);
  }

  return ThresholdState(next);
}

int calculateAlarm(TelemetryState telemetry, ThresholdState thresholds) {
  final distance = telemetry['distance_cm'];
  final threshold = thresholds['distance'];
  if (distance == null || threshold == null) {
    return 0;
  }
  return distance > 0 && distance < threshold ? 1 : 0;
}

String serializeCommand(Map<String, dynamic> command) => jsonEncode(command);

String bytesToString(Uint8List bytes) => utf8.decode(bytes);

Uint8List stringToBytes(String text) => Uint8List.fromList(utf8.encode(text));

num _toNumber(Object? value, num fallback) {
  if (value is num) {
    return value.isFinite ? value : fallback;
  }
  final parsed = num.tryParse(value?.toString() ?? '');
  return parsed == null || !parsed.isFinite ? fallback : parsed;
}

int _toSwitch(Object? value) {
  if (value is bool) {
    return value ? 1 : 0;
  }
  if (value is num) {
    return value == 0 ? 0 : 1;
  }
  final text = value?.toString().toLowerCase();
  return text == null || text == '0' || text == 'false' || text.isEmpty ? 0 : 1;
}

File: projects/DCLD_001/DCLD_001_upper_ui/src/services/dcldProtocol.js
import { getSupportedMetrics, getSupportedThresholds } from '@/config/versionCapabilities'

export function buildGetStatusCommand() {
  return { cmd: 'get_status' }
}

export function buildSetModeCommand(mode) {
  return { cmd: 'set_mode', mode }
}

export function buildSetThresholdCommand(value) {
  return { cmd: 'set_threshold', value: Number(value) }
}

export function createDefaultTelemetry(features) {
  const telemetry = {
    distance_cm: 0,
    raw_distance_cm: 0,
    temperature_c: features.temp ? 20 : null,
    threshold_cm: 80,
    mode: 'auto',
    alarm: 0,
    camera_url: '',
  }

  for (const metric of getSupportedMetrics(features)) {
    if (telemetry[metric.key] === undefined) {
      telemetry[metric.key] = null
    }
  }

  return telemetry
}

export function createDefaultThresholds() {
  return getSupportedThresholds().reduce((acc, threshold) => {
    acc[threshold.key] = threshold.defaultThreshold
    return acc
  }, {})
}

function toNumber(value, fallback = 0) {
  const number = Number(value)
  return Number.isFinite(number) ? number : fallback
}

function toSwitch(value) {
  return value ? 1 : 0
}

function normalizeMode(mode) {
  return ['auto', 'threshold'].includes(mode) ? mode : 'auto'
}

export function parseTelemetry(payload) {
  if (payload == null) {
    return {}
  }
  if (typeof payload === 'object' && !(payload instanceof ArrayBuffer)) {
    return payload
  }
  const text = payload instanceof ArrayBuffer ? arrayBufferToString(payload) : String(payload)
  try {
    return JSON.parse(text)
  } catch (error) {
    return { raw: text }
  }
}

export function normalizeTelemetry(payload, features, previous = createDefaultTelemetry(features), thresholds = createDefaultThresholds()) {
  const raw = parseTelemetry(payload)
  const next = { ...previous }

  for (const metric of getSupportedMetrics(features)) {
    if (raw[metric.key] !== undefined) {
      next[metric.key] = toNumber(raw[metric.key], previous[metric.key] || 0)
    }
  }

  if (raw.distance_cm !== undefined) {
    next.distance_cm = toNumber(raw.distance_cm, previous.distance_cm || 0)
  }
  if (raw.raw_distance_cm !== undefined) {
    next.raw_distance_cm = toNumber(raw.raw_distance_cm, previous.raw_distance_cm || 0)
  }
  if (raw.temperature_c !== undefined) {
    next.temperature_c = toNumber(raw.temperature_c, previous.temperature_c || 20)
  }
  if (raw.threshold_cm !== undefined) {
    next.threshold_cm = toNumber(raw.threshold_cm, previous.threshold_cm || 80)
  }
  if (raw.camera_url !== undefined) {
    next.camera_url = String(raw.camera_url || '')
  }
  if (raw.mode !== undefined) {
    next.mode = normalizeMode(raw.mode)
  }

  next.alarm = raw.alarm !== undefined ? toSwitch(raw.alarm) : calculateAlarm(next, thresholds)
  return { telemetry: next, raw }
}

export function normalizeThresholds(payload, previous = createDefaultThresholds()) {
  const raw = parseTelemetry(payload)
  const source = raw.thresholds && typeof raw.thresholds === 'object' ? raw.thresholds : raw
  const next = { ...previous }

  if (source.distance !== undefined) {
    next.distance = toNumber(source.distance, previous.distance ?? 80)
  }
  if (raw.threshold_cm !== undefined) {
    next.distance = toNumber(raw.threshold_cm, previous.distance ?? 80)
  }

  return next
}

export function calculateAlarm(telemetry, thresholds) {
  const distance = Number(telemetry.distance_cm)
  const threshold = Number(thresholds.distance)
  return Number.isFinite(distance) && Number.isFinite(threshold) && distance > 0 && distance < threshold ? 1 : 0
}

export function serializeCommand(command) {
  return JSON.stringify(command)
}

export function arrayBufferToString(buffer) {
  if (typeof TextDecoder !== 'undefined') {
    return new TextDecoder('utf-8').decode(buffer)
  }
  const bytes = new Uint8Array(buffer)
  return Array.from(bytes).map((item) => String.fromCharCode(item)).join('')
}

export function stringToArrayBuffer(text) {
  if (typeof TextEncoder !== 'undefined') {
    return new TextEncoder().encode(text).buffer
  }
  const buffer = new ArrayBuffer(text.length)
  const bytes = new Uint8Array(buffer)
  for (let i = 0; i < text.length; i += 1) {
    bytes[i] = text.charCodeAt(i) & 0xff
  }
  return buffer
}
